### ISSUE TITLE
Slow reorganizations/reparses

### ISSUE THREAD

**Issue by adamkrellenstein** [CounterpartyXCP/counterparty-lib#763]
* [ ] Possible mitigation: disabling 'server behind backend' API error during a reorg.

**Comment by ouziel-slama** [comment]
About the reparse, I am still searching but I'am not sure there is a way to speed up the parsing except by trying to optimize each message parsing (for instance `get_holders` for `dividend`).

About the reorganization, the first idea that comes to my mind is to not reparse when there is a reorg, but instead just make a rollback. Perhaps, as the asset conservation checking, we can make it optional.
For the possible mitigation, it's doable IMHO if we return with each result the status of the server/backend (last backend block, last parsed block, version).

**Comment by adamkrellenstein** [comment]
A 'rollback' requires a full reparse, actually. That's the problem. :/

**Comment by ouziel-slama** [comment]
Yeah! instead to reparse every blocks, we can:
- clean every tables by using the field `block_index`
- erase and regenerate the table `balances` by "parsing" the table `credits` and `debits`.
- continue the parsing

**Comment by ouziel-slama** [comment]
For the second step it's possible to regenerate the `balances` table from `credits` and `debits` with one SQL query:

```
INSERT INTO balances (address, asset, quantity)
    SELECT address, asset, quantity FROM (
        SELECT (address || asset) AS aa, address, asset, SUM(q) AS quantity FROM (
            SELECT (address || asset) AS aa, address, asset, SUM(quantity) AS q FROM credits GROUP BY aa
            UNION ALL
            SELECT (address || asset) AS aa, address, asset, -1 * SUM(quantity) AS q FROM debits GROUP BY aa
        ) GROUP BY aa
    );
```

**Comment by adamkrellenstein** [comment]
I think the problem with that idea is rolling back complicated things like orders, whose state can change after creation.

**Comment by ouziel-slama** [comment]
yeah I realised that.. to works we should never use 'UPDATE'.. seems possible but with a big bunch of work and tests..

**Comment by adamkrellenstein** [comment]
Yeah, not worth it right now. :/

**Comment by adamkrellenstein** [comment]
We could also do 'snapshots' of the database by just copying the DB files? @ouziel-slama?

**Comment by ouziel-slama** [comment]
yeah, I thought about that, but how many snapshots we should keep ? this system can consume a lot of space. Ideally we need a kind of "incremental snapshot", something a la git.. why not git itself ?

**Comment by ouziel-slama** [comment]
In the same spirit as snapshots, we can use the `exectracer` to keep an history of every SQL queries, and "replay" queries until the desired block.

**Comment by adamkrellenstein** [comment]
Two snapshots, at say 15 and 30 blocks back? Reorgs never go back more than 10 blocks.

Git isn't good with large files :/

It all seems so hackish! ;) Shouldn't this be a solved problem??

**Comment by ouziel-slama** [comment]
:-) yeah I agree! the cleanest way is to completely avoid 'UPDATE' and so to be able "prune" cleanly the database but I agree it's not worth right now.

**Comment by deweller** [comment]
This is becoming a significant problem for us now.

Counterparty took over 50 minutes to reparse all transactions today at block 363360.

```
2015-07-01-T15:11:34+0000 [WARNING] Blockchain reorganisation at block 363360.
2015-07-01-T15:11:34+0000 [INFO] Reparsing all transactions.
2015-07-01-T15:57:07+0000 [INFO] Database minor version number updated.
2015-07-01-T16:02:55+0000 [INFO] Issuance: 1AYS3rLqZZVEx2e7kKuHdSE2S73LHsL9Wg created 1000000000.0 of divisible asset DBSCOIN (8b1bf0fedddfd8bac8637b51cc3316fa646bbf47935978ddcf102a8e3337ea86) [valid]
2015-07-01-T16:02:55+0000 [INFO] Block: 363360 (1.32s)
```

**Comment by deweller** [comment]
Here is a log from @jdog:

```
2015-07-01-T14:57:09+0000 [INFO] Block: 363360 (2.38s)
2015-07-01-T15:11:42+0000 [WARNING] Blockchain reorganisation at block 363360.
2015-07-01-T15:11:42+0000 [INFO] Reparsing all transactions.
2015-07-01-T15:29:29+0000 [INFO] Database minor version number updated.
2015-07-01-T15:29:38+0000 [INFO] Issuance: 1AYS3rLqZZVEx2e7kKuHdSE2S73LHsL9Wg created 1000000000.0 of divisible asset DBSCOIN (8b1bf0fedddfd8bac8637b51cc3316fa646bbf47935978ddcf102a8e3337ea86) [valid]
2015-07-01-T15:29:38+0000 [INFO] Block: 363360 (1.21s)
```

He is using a much higher powered machine and his reparse took 18 minutes.

**Comment by deweller** [comment]
Here is a proposal for a solution.  

For every database update, save the change in an `undo` table.  The purpose of the undo table is to unwind every database change applied during a block.  

Perhaps the undo table for `credits` might look like this:

```
CREATE TABLE IF NOT EXISTS credits_undo(
undo_serial INTEGER,
undo_type TEXT,
undo_fields TEXT,
block_index INTEGER,

address TEXT,
asset TEXT,
quantity INTEGER,
calling_function TEXT,
event TEXT,
)
```

The undo fields have a special purpose.

+ `undo_serial` is an incrementing, unique id.
+ `undo_type` is CREATE, UPDATE, or DELETE
+ `undo_fields` is a list of fields that were modified like `quantity,calling_function,event`.  (This may not be needed.)

For UPDATE and DELETE events, the rest of the fields contain the values that were present in the database row before the UPDATE or DELETE query was applied.  For CREATE events, the rest of the fields can be left empty.


To unwind the last 10 blocks after a blockchain reorganization, iterate through all the undo table entries starting with the highest `undo_serial` and work backwords.  Using the information in the undo table, a special process could undo each database query.  This special undo code will restore each table back to its original state before the database query was applied to the table.


On a successful block parse, we can prune all the undo tables, so that we are only keeping the last 10 blocks worth of changes.


Questions:

+ Are there large update or delete queries in use now that make this infeasible?

+ Would this approach require that all tables have unique IDs?  If so, is it feasible to add them?

**Comment by deweller** [comment]
I just discovered that sqlite has a built in unique ID for every row called `rowid`. (Who knew?)  So that answers the second question above.

You could also implement the undo log at the database level which would make things cleaner and require a lot less python code.  Here is a proof of concept example from some old SQLIte Docs:

http://www.sqlite.org/cvstrac/wiki?p=UndoRedo

**Comment by adamkrellenstein** [comment]
I'll (optionally) disable the `'behind backend`' error during reorgs for now.

Also, I'll make sure that `get_running_info` accurately displays the state of the db.

**Comment by adamkrellenstein** [comment]
See #782.

**Comment by robby-dermody** [comment]
@deweller: have you seen the messages log table? (https://github.com/CounterpartyXCP/counterparty-lib/blob/master/counterpartylib/lib/blocks.py#L327). perhaps that will work as the basis for undo operations?

**Comment by robby-dermody** [comment]
@deweller I talked a bit to @adamkrellenstein about this... theoretically an approach like an undo log should work. right now we're rather strapped for bandwidth. if you (or anyone else reading this) would like to make an attempt at a proof of concept, we'd be happy to chat through an exact approach. just let us know

**Comment by deweller** [comment]
Thanks.  We really want this to happen at Tokenly.  We are discussing finding a way to make this happen.  We'll let you know any approach we plan to take before we proceed on developing a patch.

**Comment by robby-dermody** [comment]
@deweiller   Absolutely, once you get a green light, we can all chat about the best way to go about it. It doesn't seem super difficult, but will require a good deal of testing to ensure that it works in all rollback scenarios...

**Comment by robby-dermody** [comment]
My first attempt at this: https://github.com/CounterpartyXCP/counterparty-lib/commit/7e50c4739502cb541a0fb641dc026a2855a132dc

will be working the bugs out over the next day or two... looks like this may work...

**Comment by adamkrellenstein** [comment]
@robby-dermody :+1: 

How fast is a reorg with that system?

**Comment by robby-dermody** [comment]
Pretty much immediate... I've done my initial pass of work and weeding out bugs. After a rollback, hash is off, so I'll see if I can find where it's not matching up

**Comment by ouziel-slama** [comment]
@robby-dermody: nice!!

**Comment by robby-dermody** [comment]
https://github.com/CounterpartyXCP/counterparty-lib/commits/undolog

started with a bootstrapped CP DB, synced up to the current block. stopped the server and did a rollback 5 blocks back, which utilized the undolog (and included a number of counterparty transactions). Started it back up, had it catch back up, and the consensus hashes matched! Yay!

Will be leaving it running for a day or so to hopefully catch a reorg or two.

@deweller please check this our in your staging environment as well.

**Comment by robby-dermody** [comment]
Example undolog rollback:

```[DEBUG] Executing undolog seq 2423: DELETE FROM sends WHERE rowid=323007
[DEBUG] Executing undolog seq 2422: DELETE FROM messages WHERE rowid=1119940
[DEBUG] Executing undolog seq 2421: DELETE FROM credits WHERE rowid=370510
[DEBUG] Executing undolog seq 2420: DELETE FROM messages WHERE rowid=1119939
[DEBUG] Executing undolog seq 2419: DELETE FROM balances WHERE rowid=90628
[DEBUG] Executing undolog seq 2418: DELETE FROM debits WHERE rowid=300376
[DEBUG] Executing undolog seq 2417: DELETE FROM messages WHERE rowid=1119938
[DEBUG] Executing undolog seq 2416: UPDATE balances SET address='15EtyL7pvXdrB582j7ZjDTwVDVDTLYCv3g',asset='ETHXCPCARD',quantity=16 WHERE rowid=90396
[DEBUG] Executing undolog seq 2415: DELETE FROM order_expirations WHERE rowid=319340
[DEBUG] Executing undolog seq 2414: DELETE FROM messages WHERE rowid=1119937
[DEBUG] Executing undolog seq 2413: DELETE FROM credits WHERE rowid=370509
[DEBUG] Executing undolog seq 2412: DELETE FROM messages WHERE rowid=1119936
[DEBUG] Executing undolog seq 2411: UPDATE balances SET address='1Es1BQJvATSKiC1Hx6yXJbZ28BRMJZxn8a',asset='SCOTCOIN',quantity=495798521163416 WHERE rowid=67559
[DEBUG] Executing undolog seq 2410: DELETE FROM messages WHERE rowid=1119935
[DEBUG] Executing undolog seq 2409: UPDATE orders SET tx_index=319340,tx_hash='a5926da6a3434e92df12387581e855d594a290f770980fcec77c2ae45960f446',block_index=374690,source='1Es1BQJvATSKiC1Hx6yXJbZ28BRMJZxn8a',give_asset='SCOTCOIN',give_quantity=50000000000000,give_remaining=50000000000000,get_asset='XCP',get_quantity=4499500000,get_remaining=4499500000,expiration=998,expire_index=375688,fee_required=0,fee_required_remaining=0,fee_provided=10000,fee_provided_remaining=10000,status='open' WHERE rowid=21569
[DEBUG] Executing undolog seq 2408: DELETE FROM sends WHERE rowid=323006
[DEBUG] Executing undolog seq 2407: DELETE FROM messages WHERE rowid=1119934
[DEBUG] Executing undolog seq 2406: DELETE FROM credits WHERE rowid=370508
[DEBUG] Executing undolog seq 2405: DELETE FROM messages WHERE rowid=1119933
[DEBUG] Executing undolog seq 2404: UPDATE balances SET address='15EtyL7pvXdrB582j7ZjDTwVDVDTLYCv3g',asset='BITCRYSTALS',quantity=58500000000 WHERE rowid=90408
[DEBUG] Executing undolog seq 2403: DELETE FROM debits WHERE rowid=300375
[DEBUG] Executing undolog seq 2402: DELETE FROM messages WHERE rowid=1119932
[DEBUG] Executing undolog seq 2401: UPDATE balances SET address='1Gpe8KHgZpgzpVLwri1UnU99wwCUWmm3cs',asset='BITCRYSTALS',quantity=98000000000 WHERE rowid=89549
```

**Comment by robby-dermody** [comment]
Ok. Fixed the issue. Seems to hold consensus well after numerous rollbacks. (However, the consensus hashes only seem to be calculated from debits and credits). I did compare some diffs of table dumps between this patch and `master` and those look identical as well.

@deweller give it a try if you can. I'll continue to run for a few days. This version prints the consensus hashes to the screen. I modified `master` on a few systems to do the same, so I could compare.

If it continues to look good in a few days through a number of reorgs, I'll propose a PR to `develop`.

**Comment by adamkrellenstein** [comment]
@robby-dermody sounds good!

**Comment by robby-dermody** [comment]
Also, the tests are failing, but at a quick glance that appears to be only because of the new SQL in use (the SQL diffs fail). I think I'll need @ouziel-slama 's help to double check that and fix.

**Comment by robby-dermody** [comment]
Resolved in 9.52.0. Closing
